Lib/test/test_re.py

```diff
@@ -2101,6 +2101,14 @@ def test_bug_34294(self):
                           {'tag': 'foo', 'text': None},
                           {'tag': 'foo', 'text': None}])
 
+    def test_bug_35859(self):
+        # Capture behavior depends on the order of an alternation
+        s = 'ab'
+        self.assertEqual(re.search(r'(ab|a)*?b', s).groups(), ('a',))
```

Comment:
Why `search()` is used instead of `match()` or `fullmatch()`?

`ab|a` is equivalent to `ab?`. Is there a reason why use the former? If there is a difference, it is better to use `.b|a` instead, because `ab|a` can be transformed to `ab?` by the RE compiler in future versions.

Reply:
Nice catch, this PR doesn't fix the problem.
```
>>> re.match(r'(ab?)*?b', 'ab').groups()
('',)
```
The correct output should be:
```
>>> regex.match(r'(ab?)*?b', 'ab').groups()
('a',)
```
I will recheck the patch tomorrow.

Reply:
@serhiy-storchaka 

> Why `search()` is used instead of `match()` or `fullmatch()`?

"Because it can be". This is a reduced test case for a real regex I found during my research.